waiter/src/waiter/token_watch.clj

```diff
@@ -0,0 +1,151 @@
+                      (contains? include-flags "token-index-map")
+                      (assoc :token-index-map token-index-map)
+                      (contains? include-flags "buffer-state")
+                      (assoc :buffer-state {:update-chan-count (count tokens-update-chan-buffer)
+                                            :watch-channels-update-chan-count
+                                            (count tokens-watch-channels-update-chan-buffer)}))))
+          go-chan
+          (async/go
+            (try
+              (loop [{:keys [token-index-map watch-chans] :as current-state} @state-atom]
+                (reset! state-atom current-state)
+                (let [[msg current-chan]
+                      (async/alts! [exit-chan tokens-update-chan tokens-watch-channels-update-chan
+                                    watch-refresh-timer-chan query-chan]
+                                   :priority true)
+                      next-state
+                      (condp = current-chan
+                        exit-chan
+                        (do
+                          (log/warn "Stopping token-watch-maintainer")
+                          (when (not= :exit msg)
+                            (throw (ex-info "Stopping router-state maintainer" {:time (clock) :reason msg}))))
+
+                        tokens-update-chan
+                        (timers/start-stop-time!
+                          (metrics/waiter-timer "core" "token-watch-maintainer" "process-tokens-update")
+                          (let [{:keys [token owner]} msg
+                                token-index-entry (tkn/get-token-index kv-store token owner :refresh true)
+                                local-token-index-entry (get token-index-map token)]
+                            (cond
+                              ; There is no change detected, so no event to be reported
+                              (= token-index-entry local-token-index-entry)
+                              current-state
+                              ; If index-entry retrieved from kv-store exists then treat as UPDATE (includes token creation)
+                              (some? token-index-entry)
+                              (let [closed-chans (send-event-to-channels watch-chans (tkn/make-index-event :UPDATE token-index-entry))]
+                                (-> current-state
+                                    (assoc-in [:token-index-map token] token-index-entry)
+                                    (assoc :watch-chans (apply disj watch-chans closed-chans))))
+                              :else
+                              (let [event-obj {:owner owner :token token}
+                                    closed-chans (send-event-to-channels watch-chans (tkn/make-index-event :DELETE event-obj))]
+                                (-> current-state
+                                    (assoc :token-index-map (dissoc token-index-map token))
```

Comment:
We can use `dissoc-in` here.
https://github.com/twosigma/waiter/blob/6401cdd70fe779a48073f699286fb629d2f48476/waiter/src/waiter/util/utils.clj#L475

Reply:
`Any empty maps that result will not be present in the new structure.` in the function description, so that means the token->index-map will be nil instead of an empty map. Do we distinguish between empty maps and nil?